doc/adr/0004-bulk-replace-api-data.md

```diff
@@ -0,0 +1,84 @@
+
+How to implement the second step? Three approaches come to mind
+
+1.
+    1. Add 3 end-points to the API, `PUT /courses`, `PUT /courses/offerings` and `PUT /courses/prerequisites`. Each
+       end-point accepts a body in media format `text/csv` parses and processes the supplied data and inserts records
+       into the API database.
+    2. Provide stand-alone scripts, CLI programs or UI that will authenticate with hmpps-auth, obtain a suitable token
+       and send the csv files to the correct end-points
+2.
+    1. Provide PUT, POST and DELETE end-points as required
+       for `/courses`, `/courses/{courseId}`, `/courses/{courseId}/offerings`, `/courses/{courseId}/offerings/{offeringId}`, `/courses/{courseId}/prerequisites`
+       and `/courses/{courseId}/prerequisites/{prerequisiteId}`. The PUT and POST methods accept JSON documents
+       representing the corresponding entities to be added or updated. The DELETE methods have no need of data beyond
+       the relevant Ids.
+    2. Provide stand-alone scripts, CLI programs or UI that will take CSV files downloaded from the spreadsheet, parse
+       them and use the end-points created in step 1 above to mutate the API data to align with the data in the CSV
+       files.
+3. Write scripts (Kotlin) to massage the data in the CSV files into a set of database DML statements that can be applied
+   to the database.
+
+The first approach is considerably simpler than the second and will take less time to implement. In its simplest form
+the logic behind the API end-points would delete all existing data and then insert the new data. This is quick and easy
+to do, but is unusable in the long term because the API links courses with prerequisites and offerings using UUID
+values (id) that are created when the entity records are inserted. These ids are the primary keys of the entities and
+should not change. Replacing the entities also replaces the ids which is probably unacceptable when other data such as
+referrals, links to course data using the ids.
+
+The third approach, mutating CSV files into DML scripts (INSERT, UPDATE etc) might be even simpler than the first,
+especially if the generated DML scripts were added to the API's GitHub repository. The scripts would be applied to the
+database using the automated Flyway database migrations that are run each time the API application is started.
+Unfortunately the GitHub repository is public, programme data should not be published on the internet and must not be
+stored in the GitHub repository.
+Therefore, the DML scripts must be applied directly to the database using a database client such as `psql` or
+the PostgreSQL database support in IntelliJ. Taking this approach requires
+
+* Tunneling a database connection through the MOJ platform to an RDS database using the kubectl command-line client
+* Extracting database credentials from a Kubernetes secret to authenticate the database connection.
```

Comment:
`tunneling` and `extracting` (again these points form a sentence based on the structure of the pretext)

`database connection` (no full stop; or you could add a semicolon to the end of the first and retain the full stop, but this would need changing in the previous document/commit too for consistency)